aidial_adapter_openai/utils/tokenizer.py

```diff
@@ -0,0 +1,125 @@
+"""
+Implemented based on the official recipe: https://cookbook.openai.com/examples/how_to_count_tokens_with_tiktoken
+"""
+
+from abc import abstractmethod
+from typing import Generic, List, TypeVar
+
+from aidial_sdk.exceptions import InternalServerError
+from tiktoken import Encoding, encoding_for_model
+
+from aidial_adapter_openai.utils.image_tokenizer import tokenize_image_by_size
+from aidial_adapter_openai.utils.multi_modal_message import MultiModalMessage
+
+TokenizeInput = TypeVar("TokenizeInput")
+
+
+class BaseTokenizer(Generic[TokenizeInput]):
+    model: str
+    encoding: Encoding
+    TOKENS_PER_REQUEST = 3
+
+    def __init__(self, model: str) -> None:
+        self.model = model
+        try:
+            self.encoding = encoding_for_model(model)
+        except KeyError as e:
+            raise InternalServerError(
+                f"Could not find tokenizer for the model {model!r} in tiktoken. "
+                "Consider mapping the model to an existing tokenizer via MODEL_ALIASES env var, "
+                "or declare it as a model which doesn't require tokenization through tiktoken.",
+            ) from e
+
+    def calculate_text_tokens(self, text: str) -> int:
+        return len(self.encoding.encode(text))
+
+    @property
+    def tokens_per_message(self) -> int:
+        """
+        Tokens, that are counter for each message, regardless of its content
+        """
+        if self.model == "gpt-3.5-turbo-0301":
+            return 4
+        return 3
+
+    @property
+    def tokens_per_name(self) -> int:
+        """
+        Tokens, that are counter for "name" field in message, if it's present
+        """
+        if self.model == "gpt-3.5-turbo-0301":
+            return -1
+        return 1
+
+    def calculate_request_prompt_tokens(self, messages_tokens: int):
+        """
+        Amount of tokens, that will be counted by API
+        is greater than actual sum of tokens of all messages by PROMPT_TOKENS
```

Comment:
`PROMPT_TOKENS => TOKENS_PER_REQUEST`

Not sure we need this comment for a single line of code with very telling namings...

Reply:
I don't see any harm in leaving the comment
It still can add some context to person, that is not familiar with tokenization